### PR TITLE
Pipeline tools detection

### PR DESCRIPTION
**Pull request by guyacosta**
#308 Enhancement request to detect tools used in pipeline to help determine additional post build sdlc tools or processes needed.

### CODE REVIEW

**Comment by scovetta** [comment]
I'll take a look tonight, please hold off on merging until we sync.

File: AppInspector/rules/default/tools/pipeline.json [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    "description": "Pipeline Tool (AntiMalware)",
+    "tags": [
+      "Metadata.Pipeline.Tools.AntiMalware"
+    ],
+    "applies_to": ["yaml"],
+    "severity": "critical",
+    "patterns": [
+      {
+        "pattern": "AntiMalware",
+        "type": "regex",
```
Comment (scovetta) [inline AppInspector/rules/default/tools/pipeline.json:14]: Can just be a string.

File: AppInspector/rules/default/tools/pipeline.json [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    "description": "A tool designed to find common security issues in Python code",
+    "tags": [
+      "Metadata.Pipeline.Tools.Bandit"
+    ],
+    "applies_to": ["yaml"],
+    "severity": "moderate",
+    "patterns": [
+      {
+        "pattern": "Bandit@1",
+        "type": "regex",
```
Comment (scovetta) [inline AppInspector/rules/default/tools/pipeline.json:34]: Can just be a string. (I'll avoid comment spamming, but a lot of these should just be turned into strings.)

File: AppInspector/rules/default/tools/pipeline.json [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+        "modifiers": [ "i" ],
+        "confidence": "high",
+        "_comment": ""
+      }
+    ]
+  },
+  {
+    "name": "Pipeline Tool (CodeQL)",
+    "id": "AI053600",
+    "description": "Each CodeQL query represents a piece of security knowledge � codified, readable, and executable � ready to be applied to any number of projects",
```
Comment (scovetta) [inline AppInspector/rules/default/tools/pipeline.json:65]: Weird characters in the description.

Comment (guyacosta) [reply AppInspector/rules/default/tools/pipeline.json:65]: hmm copy paste issue -ok

File: AppInspector/rules/default/tools/pipeline.json (hunk 2 above)
Comment (guyacosta) [reply AppInspector/rules/default/tools/pipeline.json:34]: yes -started with a copy pattern thanks will adjust

File: UnitTest.Commands/html/index.html [hunk 4]
```diff
@@ -3,24 +3,24 @@
 <head>
     <meta charset="utf-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
-    <link rel="stylesheet" href="html/resources/css/bootstrap.min.css" />
-    <link rel="stylesheet" href="html/resources/css/all.css" />
-    <link rel="stylesheet" href="html/resources/css/c3.min.css" />
+    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.5.2/css/bootstrap.min.css" />
```
Comment (scovetta) [inline UnitTest.Commands/html/index.html:6]: External links to to JS or CSS should have SRI attributes, like:
```
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.5.3/css/bootstrap.min.css" integrity="sha512-oc9+XSs1H243/FRN9Rw62Fn8EtxjEYWHXRvjS43YtueEewbS6ObfXcJNyohjHqVKFPoXXUxwc+q1K7Dee6vv9g==" crossorigin="anonymous" />
```

You can pick up those links right from the [cdnjs](https://cdnjs.com/libraries/twitter-bootstrap) page.

File: UnitTest.Commands/html/index.html [hunk 5]
```diff
@@ -51,16 +51,16 @@
...
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/ace/1.4.12/ace.js" integrity="sha512-GZ1RIgZaSc8rnco/8CXfRdCpDxRCphenIiZ2ztLy3XQfCbQUSCuk8IudvNHxkRA3oUg6q0qejgN/qqyG1duv5Q==" crossorigin="anonymous"></script>
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.5.2/js/bootstrap.min.js" integrity="sha512-M5KW3ztuIICmVIhjSqXe01oV2bpe248gOxqmlcYrEzAvws7Pw3z6BK0iGbrwvdrUQUhi3eXgtxp5I8PDo9YfjQ==" crossorigin="anonymous"></script>
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/c3/0.7.20/c3.min.js" integrity="sha512-+IpCthlNahOuERYUSnKFjzjdKXIbJ/7Dd6xvUp+7bEw0Jp2dg6tluyxLs+zq9BMzZgrLv8886T4cBSqnKiVgUw==" crossorigin="anonymous"></script>
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min.js" integrity="sha512-FHsFVKQ/T1KWJDGSbrUhTJyS1ph3eRrxI228ND0EGaEp6v4a/vGwPWd3Dtd/+9cI7ccofZvl/wulICEurHN1pg==" crossorigin="anonymous"></script>
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/2.5.3/umd/popper.min.js"></script>
```
Comment (scovetta) [inline UnitTest.Commands/html/index.html:59]: Same SRI comment as above.

**Comment by guyacosta** [comment]
Only the AppInspector/rules/default/tools/pipeline.json needs to be reviewed.  The other previous changes are already merged into main so not clear why they are included in this PR.

File: UnitTest.Commands/html/index.html (hunk 4 above)
Comment (guyacosta) [reply UnitTest.Commands/html/index.html:6]: This isn't right.  My local copy does have the integrity values.  I'm unclear why there is a difference.

File: UnitTest.Commands/html/partials/_report_summary.liquid [hunk 6]
```diff
@@ -86,36 +86,36 @@
                 </tr>
                 <tr>
                   <td style="width:300px;">Application Types</td>
-                  <td style="">{% for item in apptypes -%}{{item}} {% endfor -%}</td>
+                  <td style="">{% for item in apptypes -%}{{item}}, {% endfor -%}</td>
```
Comment (scovetta) [inline UnitTest.Commands/html/partials/_report_summary.liquid:89]: These will have a comma appended to it. I think you can use the 'join' template tag:

```
{{ apptypes | join:", " }}
```

File: UnitTest.Commands/html/index.html (hunk 4 above)
Comment (guyacosta) [reply UnitTest.Commands/html/index.html:6]: See how the merged version actually has the where it was available in the already merged and previous branch so again not clear how these are incorrect. 
 Compare https://github.com/microsoft/ApplicationInspector/blob/main/UnitTest.Commands/html/index.html

**Comment by guyacosta** [comment]
This PR has incorrect changes in the commit.  Will resubmit new with relevant suggestions.

**Pull request closed without merging**